File: src/main/java/net/cjsah/console/command/Dispatcher.java
package net.cjsah.console.command;

import com.google.common.collect.Lists;
import net.cjsah.console.command.builder.LiteralArgumentBuilder;
import net.cjsah.console.command.context.CommandContext;
import net.cjsah.console.command.context.CommandContextBuilder;
import net.cjsah.console.command.source.CommandSource;
import net.cjsah.console.command.tree.CommandNode;
import net.cjsah.console.command.tree.RootCommandNode;
import net.cjsah.console.exceptions.BuiltExceptions;
import net.cjsah.console.exceptions.CommandException;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class Dispatcher {
    private static final char ARGUMENT_SEPARATOR = ' ';

    private static final RootCommandNode ROOTS = new RootCommandNode();

    public void register(LiteralArgumentBuilder command) {
        ROOTS.addChild(command.build());
    }

    protected int execute(String input, CommandSource<?> source) throws CommandException {
        StringReader reader = new StringReader(input);
        CommandContextBuilder builder = new CommandContextBuilder(this, source, ROOTS, reader.getCursor());
        ParseResults parse = parseNodes(ROOTS, reader, builder);
        if (reader.canRead()) {
            if (parse.getExceptions().size() == 1) throw parse.getExceptions().values().iterator().next();
            else if (parse.getContext().getRange().isEmpty()) throw BuiltExceptions.dispatcherUnknownCommand().createWithContext(reader);
            else throw BuiltExceptions.dispatcherUnknownArgument().createWithContext(reader);
        }
        int result = 0;
        boolean foundCommand = false;
        CommandContext original = parse.getContext().build();
        List<CommandContext> contexts = Lists.newArrayList(original);
        while (contexts != null) {
            for (CommandContext context : contexts) {
                if (context.getCommand() != null) {
                    foundCommand = true;
                    try {
                        context.getCommand().run(context);
                        result += 1;
                    } catch (CommandException e) {
                        e.printStackTrace();
                    }
                }
            }
            contexts = null;
        }
        if (!foundCommand) {
            throw BuiltExceptions.dispatcherUnknownCommand().createWithContext(reader);
        }
        return result;
    }

    private ParseResults parseNodes(CommandNode node, StringReader originalReader, CommandContextBuilder builder) {
        Map<CommandNode, CommandException> exceptions = new LinkedHashMap<>();
        CommandSource<?> source = builder.getSource();
        List<ParseResults> potentials = new ArrayList<>();
        int cursor = originalReader.getCursor();
        for (CommandNode child : node.getRelevantNodes(originalReader)) {
            if (!child.canUse(source)) {
                continue;
            }
            CommandContextBuilder context = builder.copy();
            StringReader reader = originalReader.copy();
            try {
                try {
                    child.parse(reader, context);
                }catch (RuntimeException e) {
                    throw BuiltExceptions.dispatcherParseException().createWithContext(reader, e.getLocalizedMessage());
                }
                if (reader.canRead() && reader.peek() != ARGUMENT_SEPARATOR) throw BuiltExceptions.dispatcherExpectedArgumentSeparator().createWithContext(reader);
            } catch (CommandException e) {
                exceptions.put(child, e);
                reader.setCursor(cursor);
                continue;
            }
            context.withCommand(child.getCommand());
            if (reader.canRead(2)) {
                reader.skip();
                ParseResults parse = parseNodes(child, reader, context);
                potentials.add(parse);
            } else {
                potentials.add(new ParseResults(context, reader, Collections.emptyMap()));
            }
        }
        if (!potentials.isEmpty()) {
            if (potentials.size() > 1) {
                potentials.sort((a, b) -> {
                    if (!a.getReader().canRead() && b.getReader().canRead()) {
                        return -1;
                    }
                    if (a.getReader().canRead() && !b.getReader().canRead()) {
                        return 1;
                    }
                    if (a.getExceptions().isEmpty() && !b.getExceptions().isEmpty()) {
                        return -1;
                    }
                    if (!a.getExceptions().isEmpty() && b.getExceptions().isEmpty()) {
                        return 1;
                    }
                    return 0;
                });
            }
            return potentials.get(0);
        }
        return new ParseResults(builder, originalReader, exceptions);
    }

    public Map<String, String> getHelp(CommandSource<?> source) {
        final Map<String, String> result = new LinkedHashMap<>();
        ROOTS.getChildren().forEach((child) -> this.getHelp(result, "/", child, source, true));
        return result;
    }

    private void getHelp(final Map<String, String> result, String mem, CommandNode node, CommandSource<?> source, boolean first) {
        if (!node.canUse(source)) return;
        mem += String.format("%s%s", first ? "" : " ", node.getUsageText());
        if (node.getCommand() != null) {
            result.put(mem, node.getHelp());
        }
        String finalMem = mem;
        node.getChildren().forEach((child) -> this.getHelp(result, finalMem, child, source, false));
    }
}

File: src/main/kotlin/net/cjsah/console/command/Dispatcher.kt
@file:Suppress("PrivatePropertyName", "unused")
@file:JvmBlockingBridge

package net.cjsah.console.command

import me.him188.kotlin.jvm.blocking.bridge.JvmBlockingBridge
import net.cjsah.console.Console
import net.cjsah.console.command.builder.LiteralArgumentBuilder
import net.cjsah.console.command.context.CommandContext
import net.cjsah.console.command.context.ContextBuilder
import net.cjsah.console.command.source.CommandSource
import net.cjsah.console.command.tree.CommandNode
import net.cjsah.console.command.tree.RootCommandNode
import net.cjsah.console.exceptions.BuiltExceptions
import net.cjsah.console.exceptions.CommandException
import net.cjsah.console.exceptions.ConsoleException
import net.cjsah.console.exceptions.PluginException
import net.cjsah.console.plugin.Plugin
import net.cjsah.console.text.TranslateText

class Dispatcher {
    companion object {
        private const val ARGUMENT_SEPARATOR = ' '
    }
    private val ROOTS = RootCommandNode()

    fun register(command: LiteralArgumentBuilder){
        if (Console.isFrozen())
            throw ConsoleException.create(TranslateText("command.reg.freezed"), PluginException::class.java)
        ROOTS.addChild(command.build())
    }

    fun register(plugin: Plugin, command: LiteralArgumentBuilder) = ROOTS.addChild(plugin, command.build())

    fun deregister(plugin: Plugin) = ROOTS.removeChild(plugin)

    @Throws(CommandException::class)
    suspend fun execute(input: String, source: CommandSource<*>): Int {
        val reader = StringReader(input)
        val builder = ContextBuilder(this, source, reader.getCursor())
        val parse = parseNodes(ROOTS, reader, builder)
        if (parse.reader.canRead()) {
            if (parse.exceptions.size == 1) throw parse.exceptions.values.iterator().next()
            else if (parse.context.getRange().isEmpty())
                throw BuiltExceptions.DISPATCHER_UNKNOWN_COMMAND.createWithContext(reader)
            else throw BuiltExceptions.DISPATCHER_UNKNOWN_ARGUMENT.createWithContext(reader)
        }
        var result = 0
        var foundCommand = false
        val original = parse.context.build()
        var contexts: List<CommandContext>? = listOf(original)
        while (contexts != null) {
            contexts.forEach {
                it.getCommand()?.let { command ->
                    foundCommand = true
                    try {
                        result += command.run(it)
                    } catch (e: CommandException) {
                        e.printStackTrace()
                    }
                }
            }
            contexts = null
        }
        if (!foundCommand) throw BuiltExceptions.DISPATCHER_UNKNOWN_COMMAND.createWithContext(reader)
        return result
    }

    private fun parseNodes(node: CommandNode, originalReader: StringReader, builder: ContextBuilder): ParseResults {
        val exceptions: MutableMap<CommandNode, CommandException> = LinkedHashMap()
        val source = builder.getSource()
        val potentials: MutableList<ParseResults> = ArrayList()
        val cursor = originalReader.getCursor()
        for (child in node.getRelevantNodes(originalReader)) {
            if (!child.canUse(source)) continue
            val context = builder.copy()
            val reader = originalReader.copy()
            try {
                try {
                    child.parse(reader, context)
                } catch (e: CommandException) {
                    throw BuiltExceptions.DISPATCHER_PARSE_EXCEPTION.createWithContext(reader, e.localizedMessage)
                }
                if (reader.canRead() && reader.peek() != ARGUMENT_SEPARATOR)
                    throw BuiltExceptions.DISPATCHER_EXPECTED_ARGUMENT_SEPARATOR.createWithContext(reader)
            } catch (e: CommandException) {
                exceptions[child] = e
                reader.setCursor(cursor)
                continue
            }
            context.withCommand(child.getCommand())
            if (reader.canRead(2)) {
                reader.skip()
                val parse = parseNodes(child, reader, context)
                potentials.add(parse)
            } else {
                potentials.add(ParseResults(context, reader, emptyMap()))
            }
        }
        if (potentials.isNotEmpty()) {
            if (potentials.size > 1) {
                potentials.sortWith(Comparator { a: ParseResults, b: ParseResults ->
                    if (!a.reader.canRead() && b.reader.canRead()) {
                        return@Comparator -1
                    }
                    if (a.reader.canRead() && !b.reader.canRead()) {
                        return@Comparator 1
                    }
                    if (a.exceptions.isEmpty() && b.exceptions.isNotEmpty()) {
                        return@Comparator -1
                    }
                    if (a.exceptions.isNotEmpty() && b.exceptions.isEmpty()) {
                        return@Comparator 1
                    }
                    0
                })
            }
            return potentials[0]
        }
        return ParseResults(builder, originalReader, exceptions)
    }

    fun getHelp(source: CommandSource<*>): Map<String, String> {
        val result: MutableMap<String, String> = LinkedHashMap()
        ROOTS.getChildren().forEach { this.getHelp(result, "/", it, source, true) }
        return result
    }

    private fun getHelp(result: MutableMap<String, String>, prefix: String, node: CommandNode, source: CommandSource<*>, first: Boolean) {
        var command = prefix
        if (!node.canUse(source)) return
        command += "${if (first) "" else " "}${node.getUsageText()}"
        if (node.getCommand() != null) {
            result[command] = node.getHelp()
        }
        node.getChildren().forEach { this.getHelp(result, command, it, source, false) }
    }

}